src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -511,14 +533,77 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
                     set.Free();
                     return false;
             }
-            
+
             // very rare case. 
             // some implemented interfaces are related
             // will have to instantiate interfaces and check
             hasRelatedInterfaces:
             return type.InterfacesNoUseSiteDiagnostics(basesBeingResolved).HasDuplicates(TypeSymbol.EqualsIgnoringDynamicAndTupleNamesComparer);
         }
 
+        internal static bool CheckTupleNamesAgainstBase(TypeSymbol type, Location location, DiagnosticBag diagnostics)
+        {
+            ArrayBuilder<DiagnosticInfo> builder = null;
+
+            bool result = CheckTupleNamesAgainstBase(type, ref builder);
+
+            if (builder != null)
+            {
+                foreach (var diagnosticInfo in builder)
+                {
+                    diagnostics.Add(new CSDiagnostic(diagnosticInfo, location));
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Check interfaces for tuple name differences with AllInterfaces from base type.
+        /// Returns false if there was a problem.
+        /// </summary>
+        internal static bool CheckTupleNamesAgainstBase(TypeSymbol type, ref ArrayBuilder<DiagnosticInfo> builder)
+        {
+            NamedTypeSymbol baseType = type.BaseTypeNoUseSiteDiagnostics;
+            if ((object)baseType == null)
+            {
+                return true;
+            }
+
+            var interfaces = type.InterfacesAndTheirBaseInterfacesNoUseSiteDiagnostics;
+
+            var baseInterfaces = new Dictionary<NamedTypeSymbol, NamedTypeSymbol>(TypeSymbol.EqualsIgnoringComparer.InstanceIgnoringTupleNames);
+            foreach (var baseInterface in baseType.AllInterfacesNoUseSiteDiagnostics)
+            {
+                if (!baseInterfaces.ContainsKey(baseInterface))
+                {
+                    baseInterfaces.Add(baseInterface, baseInterface);
+                    // Note: because AllInterfaces is sorted, if somehow conflicting tuple names already exist in baseInterfaces
+                    //       (for instance in specially-crafted metadata), then only the most derived re-implementation matters
+                }
+            }
+
+            bool result = true;
+            foreach (var @interface in interfaces)
+            {
+                if (baseInterfaces.TryGetValue(@interface, out NamedTypeSymbol found))
+                {
+                    if (!@interface.Equals(found)) // tuple names differ
+                    {
+                        if (builder == null)
+                        {
+                            builder = new ArrayBuilder<DiagnosticInfo>();
```

Comment:
This is not applicable now that I re-use the builder of TypeParameterDiagnosticInfos.

---
In reply to: [92049030](https://github.com/dotnet/roslyn/pull/15809#discussion_r92049030) [](ancestors = 92049030)